### PR TITLE
Set up dependabot configuration for pip

### LINKED ISSUE #488: Dependabot for more stable CI
Our CI has been incredibly unstable lately. Every time a new version of a dependency is released, something in our tests breaks, especially mypy.

Using [dependabot](https://docs.github.com/en/code-security/dependabot), we can pin all of our dependencies to a specific version. The bot will then periodically check for updates and open a PR to update the dependency version. That way, only the version update PR will break, not everyone else's PRs.

Side note: a lot of our deps don't yet have wheels for Python 3.10, or have never had wheels for Windows. I was thinking about switching from pip to conda for all CI. Unfortunately, it looks like dependabot only supports pip, not conda.

### PR DESCRIPTION
**Pull request by weiji14**
Enable dependabot to manage python dependencies pins for a more stable CI.

References:
- https://docs.github.com/github/administering-a-repository/configuration-options-for-dependency-updates

Closes #488

### CODE REVIEW

**Comment by adamjstewart** [comment]
Pushed a commit so that dependabot actually has something to update (`requirements.txt`) and we actually use it in CI. Remaining TODOs:

- [ ] Solve version compatibility issues (latest version of numpy doesn't support oldest Python)
- [ ] Make sure docs still render correctly (pytorch_sphinx_theme is very finicky about how it is installed: https://github.com/pytorch/pytorch_sphinx_theme/issues/143)
- [ ] Figure out how to use `requirements.txt` to control version of deps installed by conda
- [ ] Figure out how to only install a subset of deps for different tests (#553)

**Comment by adamjstewart** [comment]
I think this is ready for review. Some caveats to be aware of:

1. `pytorch-sphinx-theme` has no official releases, so we're at the whims of the master branch. It also can't be specified anywhere other than `docs/requirements.txt` so that's fun: https://github.com/pytorch/pytorch_sphinx_theme/issues/143. We should consider switching to `trojanzoo_sphinx_theme`, but we can do that in a separate PR.
2. I haven't figured out an easy way to control the conda-installed packages with dependabot. For now, we can just manually update `.github/workflows/tests.yaml` every time `fiona` or `rasterio` has a new release. We could try to call grep on `requirements.txt` and use the output to specify the version for conda to install, but honestly it might be easier and more worthwhile to just add wheels to [fiona](https://github.com/Toblerity/Fiona/discussions/1073) and [rasterio](https://github.com/rasterio/rasterio/discussions/2406).
3. [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html) is fun. If we want to continue to support Python 3.7 (I do), then we need to specify Python version-specific versions for all dependencies that subscribe to NEP-29's philosophy. I have no idea how dependabot will handle these. We can ignore them if we need to.
4. I don't know how this will interact with #544. I think dependabot will continue to update versions and ignore `setup.cfg`, so we have to be careful to update that ourselves whenever a new major version is released.
5. All tests now install all dependencies. This is unfortunate because I really liked the idea of #553. There are [open issues](https://github.com/dependabot/dependabot-core/issues/3940) requesting support for multiple `requirements.txt` files, but no progress yet. I wonder if poetry would solve any of these issues?

TL;DR: This isn't perfect, but I think it will solve most of our frequent CI issues. We should merge this and see how annoying it is. If it's more trouble than it's worth, we can revert it or see if poetry makes things easier.

**Comment by calebrob6** [comment]
This all seems fine to me. Why run black on notebooks here?

**Review by calebrob6 — APPROVED** [review]

**Comment by adamjstewart** [comment]
For some reason black was being installed with jupyter support by default now. I could figure out how to disable it, but I figured it would be easier to just run it on our notebooks.

**Comment by adamjstewart** [comment]
Hmm, why is dependabot listed as a TorchGeo contributor? How do I remove him before he becomes the primary developer of TorchGeo? 😆
